features/310-prerender-gear/prerender.md

```diff
@@ -0,0 +1,105 @@
+# Prerender
+
+Prerender is a gear that caches and serves pre-rendered html.
+
+## Use case
+
+Prerender's main use case is to optimize response time for single-page-application (SPA) when search engine asks skygear to provide an html, aka search-engine-optimization (SEO).
+
+## Terminology
+
+- **Prerendered html** - An html document that is the result of running some or all of an html's javascript and without any script tag. By definition, an html without any javascript is a prerendered html.
+
+- **Prerendered html cache** - A volatile or persistent cache that stores prerendered html.
+
+## Overview
+
+### Making a prerender request
+
+When skygear's gateway receives a request, before redirecting the traffic to a specific handler (e.g. cloud function), it would inspect `user-agent` to see if the request comes from search engines by matching `user-agent` against a configurable list of search engine. If it is, it asks prerender for a prerendered html by forwarding the url, otherwise it redirects the request to a specific handler as-is.
+
+Prerender handles any incoming request by first checking whether there is any cached DOM for the forwarded url. It there is a cache hit, prerender "hits the fast path" and returns the cached DOM as-is. Otherwise it proceeds to the slow path.
+    
+The slow path begins by first navigating to the forwarded url using a headless browser, aloowing the browser to proceed according to redirect transparently, then inspect the final response once it settles down to check if the url locates an html document (i.e. `content-type` is `text/html`). If not, prerender returns the response body and headers directly, otherwise it proceeds to capture the response body (the html DOM) and the response header.
+
+To allow scripts to run to its completion or to allow DOM to transform to a specific state, prerender waits for a configurable period of time before capturing the resultant DOM. The resultant DOM will have all of its script tags removed, and a `<base>` tag is inserted to the DOM appropriately to fix relative links.
```

Comment:
No, is that even a standard thing?

Reply:
Of course not a standard thing but nice-to-have feature. I am not asking for supporting this in the MVP, but is it hard to support this feature in the future?

Reply:
I think not, but since I think I have to start on another project, is it ok that I only write about it in `md`? I can leave notes in implementation to guide how to do it.

Reply:
Thanks!